pkg/volume/scaleio/sio_util_test.go

```diff
@@ -0,0 +1,252 @@
+		t.Error("Unexpected system value")
+	}
+	if data[confKey.protectionDomain] != "default" {
+		t.Error("Unexpected protection domain value")
+	}
+	if data[confKey.storagePool] != "default" {
+		t.Error("Unexpected storage pool value")
+	}
+	if data[confKey.volumeName] != "sio-vol" {
+		t.Error("Unexpected volume name value")
+	}
+	if data[confKey.fsType] != "xfs" {
+		t.Error("Unexpected fstype value")
+	}
+	if data[confKey.storageMode] != "ThinProvisioned" {
+		t.Error("Unexpected storage mode value")
+	}
+	if data[confKey.secretRef] != "test-secret" {
+		t.Error("Unexpected secret ref value")
+	}
+	if data[confKey.sslEnabled] != "false" {
+		t.Error("Unexpected sslEnabled value")
+	}
+	if data[confKey.readOnly] != "true" {
+		t.Error("Unexpected readOnly value: ", data[confKey.readOnly])
+	}
+}
+
+func TestUtilDefaultString(t *testing.T) {
+	if defaultString("", "foo") != "foo" {
+		t.Error("Unexpected value for default value")
+	}
+}
+
+func TestUtilSaveConfigs(t *testing.T) {
+	plugMgr, tmpDir := newPluginMgr(t)
+	defer os.RemoveAll(tmpDir)
+
+	plug, err := plugMgr.FindPluginByName(sioPluginName)
+	if err != nil {
+		t.Errorf("Can't find the plugin %v", sioPluginName)
+	}
+	sioPlug, ok := plug.(*sioPlugin)
+	if !ok {
+		t.Errorf("Cannot assert plugin to be type sioPlugin")
+	}
+
+	if err := saveConfigMap(sioPlug, "testns", testSioVolName, config); err != nil {
+		t.Errorf("failed to saveConfigMap %v", err)
+	}
+	confMap, err := sioPlug.host.GetKubeClient().Core().ConfigMaps("testns").Get(config[confKey.volSpecName], meta.GetOptions{})
+	if err != nil {
+		t.Errorf("failed to get confMap %v", err)
+	}
+	if confMap.Data == nil {
+		t.Errorf("configMap.Data should not be nil")
+	}
+	if confMap.Data[confKey.volumeName] != config[confKey.volumeName] {
+		t.Error("Unexpected value for configMap")
+	}
+}
+
+func TestUtilSetupConfigData(t *testing.T) {
+	plugMgr, tmpDir := newPluginMgr(t)
+	defer os.RemoveAll(tmpDir)
+
+	plug, err := plugMgr.FindPluginByName(sioPluginName)
+	if err != nil {
+		t.Errorf("Can't find the plugin %v", sioPluginName)
+	}
+	sioPlug, ok := plug.(*sioPlugin)
+	if !ok {
+		t.Errorf("Cannot assert plugin to be type sioPlugin")
+	}
+
+	data := make(map[string]string)
+	for k, v := range config {
+		data[k] = v
+	}
+	if err := setupConfigData(sioPlug, "default", data); err != nil {
+		t.Errorf("failed to setupConfigData %v", err)
+	}
+	if data[confKey.username] == "" {
+		t.Errorf("failed to merge secret")
+	}
+}
+
+func TestUtilLoadConfig(t *testing.T) {
+	plugMgr, tmpDir := newPluginMgr(t)
+	defer os.RemoveAll(tmpDir)
+
+	plug, err := plugMgr.FindPluginByName(sioPluginName)
+	if err != nil {
+		t.Errorf("Can't find the plugin %v", sioPluginName)
+	}
+	sioPlug, ok := plug.(*sioPlugin)
+	if !ok {
+		t.Errorf("Cannot assert plugin to be type sioPlugin")
+	}
+
+	data := make(map[string]string)
+	for k, v := range config {
+		data[k] = v
+	}
+	if err := saveConfigMap(sioPlug, "testns", testSioVolName, data); err != nil {
```

Comment:
I sync'd with @thockin offline. He agrees that having a k8s volume plugin persist state to a config map is pretty funky. Our recommendation, if you must persist state, is to do so to a file in the plugin directory. Make sure to keep in mind forward and backwards compatibility and file cleanup!